bundle/jsky.app.ot/src/main/scala/jsky/app/ot/ags/BagsManager.scala

```diff
@@ -0,0 +1,321 @@
+          val oldEnv = ctx.targets.envOrDefault
+          val allProbes = oldEnv.getGuideEnvironment.getReferencedGuiders.asScala.toSet
+          val assignedProbes = selOpt.map(_.assignments.map(_.guideProbe)).toList.flatten
+          val unassignedProbes = allProbes -- assignedProbes
+
+          // Clear out the guide probes that no longer have a valid assignment.
+          val clearedEnv = (oldEnv /: unassignedProbes) { (curEnv, gp) =>
+            val oldGpt = curEnv.getPrimaryGuideProbeTargets(gp).asScalaOpt
+            val newGpt = oldGpt.getOrElse(GuideProbeTargets.create(gp)).withBagsTarget(GuideProbeTargets.NO_TARGET)
+            curEnv.putPrimaryGuideProbeTargets(newGpt)
+          }
+
+          // Apply the new selection.
+          val newEnv = selOpt.fold(clearedEnv)(_.applyTo(clearedEnv))
+
+          // Update the TargetEnvironment if it is different.
+          if (!bagsTargetsMatch(oldEnv, newEnv)) {
+            muteObservation(observation)
+            ctx.targets.dataObject.foreach { targetComp =>
+              targetComp.setTargetEnvironment(newEnv)
+              ctx.targets.commit()
+
+              // Update the position angle, if necessary.
+              selOpt.foreach { sel =>
+                ctx.instrument.dataObject.foreach { inst =>
+                  val deg = sel.posAngle.toDegrees
+                  val old = inst.getPosAngleDegrees
+                  if (deg != old) {
+                    inst.setPosAngleDegrees(deg)
+                    ctx.instrument.commit()
+                  }
+                }
+              }
+            }
+            unmuteObservation(observation)
+          }
+        }
+
+        val fut = strategy.select(obsCtx, OT.getMagnitudeTable)
+        fut.onComplete {
+          case Success(selOpt) =>
+            LOG.info(s"BAGS lookup for observation=${observation.getObservationID} successful. ${resultString(selOpt)}")
+            Swing.onEDT {
+              applySelection(selOpt)
+              showTpeFeatures(selOpt)
+            }
+            taskComplete(this, success = true)
+          case Failure(ex) =>
+            LOG.warning(s"BAGS lookup for observation=${observation.getObservationID} failed.")
+            taskComplete(this, success = false)
+        }
+      }
+
+
+      curCtx.foreach { obsCtx =>
+        AgsRegistrar.currentStrategy(obsCtx).foreach { strategy =>
+          if (GuideStarSupport.hasGemsComponent(observation)) {
+            gemsLookup(obsCtx)
+          } else {
+            otherLookup(obsCtx, strategy)
+          }
+        }
+      }
+    }
+  }
+
+  private class BagsThread(id: Int) extends Runnable {
+    override def run() {
+      // Get the next task to run. This blocks on the queue.
+      val task = nextTask()
+
+      // Execute the task.
+      task.foreach(_.performLookup())
+
+      // Resubmit self for execution.
+      executor.submit(this)
+    }
+  }
+
+  private val NumThreads = 5
```

Comment:
Should the number of threads be related to the CPU count?
`Runtime.getRuntime().availableProcessors()`